libfreefare/mifare_desfire.c

```diff
@@ -141,49 +145,89 @@ static ssize_t	 read_data(FreefareTag tag, uint8_t command, uint8_t file_no, off
  * The Mifare DESFire function return value which is returned at the end of the
  * response is copied at the beginning to match the PICC documentation.
  */
-#define DESFIRE_TRANSCEIVE(tag, msg, res) \
-    DESFIRE_TRANSCEIVE2 (tag, msg, __##msg##_n, res)
-/* Native DESFire APDUs will be wrapped in ISO7816-4 APDUs:
-   CAPDUs will be 5 bytes longer (CLA+P1+P2+Lc+Le)
-   RAPDUs will be 1 byte longer  (SW1 SW2 instead of 1 status byte)
+static int
+desfire_transceive(FreefareTag tag, const char *msg, size_t msg_len, char *res, size_t res_size, size_t *res_len)
+{
+	uint8_t msg_buf[MAX_CAPDU_SIZE + 5];
+	uint8_t res_buf[MAX_RAPDU_SIZE + 1];
+
+	size_t len = 5;
+	int rc;
+
+	if (!msg) {
+		errno = EINVAL;
+		return -1;
```

Comment:
`return` in the macro allows us to exit from the function that use the macro.  With this refactoring, `return` only returns from the `desfire_transceive()` function, and the calling function has to check if `desfire_transceive()` was successfull or not in order to determine if it should continue or propagate the error.  Each time.

One solution to avoid a lot of repetition would be to use `setjmp(3)` / `longjmp(3)` to have some way to "raise" and "catch" errors, but it would adds clutter to any function calling `desfire_transceive()`.  Far from ideal from my point of view.

Maybe a lighter / simpler solution would be to wrap calls to `desfire_transceive()` in a macro that only does this check, e.g.

```c
#define DESFIRE_TRANSCEIVE(...) \
  do { \
    int n;
    if ((n = desfire_transceive(...)) < 0)
      return n;
  } while (0)
```

I am pretty sure modern compilers allow us to be sure that `desfire_transceive()` is never called without checking it's return value.  Found `-Wunused-result` after a quick search but I have not tested it.

Node: name macros that have side effects with UPPER CASED names: DESFIRE_TRANSCEIVE :wink:

Reply:
Thanks for your review!
I will check and fix this.

Reply:
BTW: Didn't we want to move away from using a macro?
Yes, there will be more lines and redundant code when checking always for the return value.
On the other hand doing so will allow more flexibility for handling errors.

I would vote for keeping desfire_transceive function and add return code checking in each function.
Already now it took me some time to understand the "hidden" variables created by the BUFFER macros. Keep the code readable :)

Reply:
I don't really have an opinion here.  I think a macro is acceptable if it is not longer than a few lines.  Anything is probably better than the current code :-)

All the BUFFER* macros are also quite disgusting, feel free to refactor this hardly if you feel brave enough!